Make QueueGenericThreadsNonBlocking complete tasks on destruction.

jkelling opened the pull request:
The deleter task of BudCpu dioes not need to hold a `shared_ptr` to the queue, because the queue manages the task, not vice-versa.

Also: bundling the removal of an `std::move()`  from a `return` statement from `TaskPkg` (belonging to `ConcurrentExecPool`) in here, which I came across investigating the issue. (I think this is too small of a fix for its own PR, thugh: *do not squash this*)

Replaces https://github.com/alpaka-group/alpaka/pull/1727

Closes https://github.com/alpaka-group/alpaka/issues/1726

jkelling committed:
BufCpu::AsyncBufAlloc::deleter: rm capture of queue by task

jkelling committed:
ConcurrentExecPool: Add some moves

TaskPkg: rm std::move from return which hinders C++17 copy elision
ThreadSafeQueue: std::move task from queue.front()

jkelling commented:
@alpaka-group/alpaka-maintainers @fwyzard  It turns out, that an `alpaka::QueueGenericThreadsNonBlocking`, through `ConcurrentExecPool`, cancels  not yet running tasks in its destructor instead of synchronizing. I always thought, alpaka is modelled rather closely after CUDA, where a stream after destruction completes all tasks. This behavior is also inconsistent with that of `QueueUniformCudaHipRtNonBlocking`, which naturally inherits the behavior of CUDA streams.

I think queues should not behave differently based on the interface. user code must not be expected to decide how to work with a queue based on which backend is being used. Is this inconsistency (undocumented) policy or a bug?

jkelling commented:
https://github.com/alpaka-group/alpaka/pull/1728/commits/99198fae9174fbac884b78337105a83c412244dc changes the behavior to be consistent with `QueueUniformCudaHipRtNonBlocking` and makes this PR work as I would expect.

@alpaka-group/alpaka-maintainers  please review and discuss.

fwyzard commented on include/alpaka/queue/QueueGenericThreadsNonBlocking.hpp:
```diff
@@ -106,6 +106,15 @@ namespace alpaka
 
+        ~QueueGenericThreadsNonBlocking()
+        {
+            if(m_spQueueImpl.unique())
```
This approach doesn't seem thread-safe, and in fact `shared_ptr::unique()` has been deprecated in c++17 and removed in c++20.

jkelling replied:
Yes.. In misread this at first as, `use_count()` being deprecated.

And yes, it is indeed not save because the devices hold `weak_ptr` to the `Impl`, which I also missed at first (otherwise it would effectively be safe because, when `use_count() == 1` in this d'tor no-one could make another copy, unless someone is holding a reference to `this` while it is being destructed ...).

This is more of a quick mockup to show how it should behave. A working way would be the canary between handle and impl mentioned [here](https://github.com/alpaka-group/alpaka/issues/1726#issuecomment-1133582736), but this requires more change, so I would like discuss what the preferred behavior is first.

One could also just has `~ConcurrentExecPool` wait for all tasks to complete instead of cancelling them, but then the non-blocking-queue's d'tor would block when the ref_count reaches zero, which is not desirable. Although, currently it would also block until the currently executing task is done....

fwyzard replied:
Naively I would expect that tasks enqueued in a non-blocking queue will eventually run, even if the queue object is destroyed (_e.g._ because it goes out of scope).
As you point out this is the behaviour we get from CUDA stream: calling `cudaStreamDestroy()` will not destroy the stream until all queued up work has completed:
> In case the device is still doing work in the stream `stream` when [cudaStreamDestroy()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1gfda584f1788ca983cb21c5f4d2033a62) is called, the function will return immediately and the resources associated with `stream` will be released automatically once the device has completed all work in `stream`.

Would it work for the destructor of `ConcurrentExecPool()` to queue one last task that will wait for all work to complete and only _then_ completes the destruction of the threads and other resources associated to the pool ?

jkelling replied:
That is the basic idea, it is not so simple, though:
1. It the d'tor (be it that of `ConcurrentExecPool` or the queue) waits for the tasks to complete destruction would block, which would be like `cudaStreamDestroy()` implicitly calling `cudaStreamSynchronize()` instead of returning immediately. Thus, destruction of the queue impl (mainly `ConcurrentExecPool`) must be avoided and moved into another thread.
2. There is another problem, though: if we put a task that holds the last ref to the queue into the queue and it destructs, we get the following call stack:
   1. `~QueueImpl`
   2. `~ConcurrentExecPool`
   3. `thread::join()`
 
   where that last call is the thread joining itself, which should result in:
   ```
   terminate called after throwing an instance of 'std::system_error'
      what():  Resource deadlock avoided
   ``` 
   At the moment, I do not understand why this is not happening in the tests with the current develop version. I had a close at `memBufTest` and found, that with the queue being captured, the ref-count are actually always one higher, than I expect, so the deleter never destructs the queue, though if queue is not captured, the deleter is sometimes not called, meaning the queue's refcount went to zero before the deleter's turn in the queue.

jkelling replied:
I created a simplified version of the queue, which crashes when joining itself:
https://gist.github.com/jkelling/792de4fc15e5b44dff4cd96c21ece45a

Is there a for why this does happen in alpaka? Beside the refcount accidentally being such, that the deleter task never destructs the queue?

I also do not see a way to solve this self-join problem with a generic task. The stopper task must be something the `ConcurrentExecPool` is aware of, so it can both destruct safely (avoid self-join) and the terminate the thread without accessing `this` again.

fwyzard replied:
> 2\. where that last call is the thread joining itself, which should result in:

You can avoid this specific problem checking that `thread.get_id() != std::this_thread::get_id()` before calling `join()`.

jkelling replied:
Yes, but then how do you tell the thread's main loop to terminate? It must not check the stop flag, because that was a member of the now-destructed `ConcurrentExecPool`. It *would* check it though, if do not tell it to just return. In the current implementation, only adding this check would make things more dangerous, because you replace an exception by undefined behavior (and not in the standard's sense where mostly nothing bad happens, but a read-after-free).

jkelling replied:
Ok, I understand why this is not crashing:
https://github.com/alpaka-group/alpaka/blob/22374dbfab3596c1bdda38fe90b4b40b9f6f43a1/include/alpaka/core/ConcurrentExecPool.hpp#L504-L524
The task is held in the main loop in that shared pointer. If the queue is empty after the deleter task the executor keeps the pointer and goes into the wait, leaking the queue. In most cases in the tests. There is an "asan" test failing, when the capture is removed without any other change, because sometimes the delete does not get executed and the buffer is leaked. The same address sanitizer does not detect the leak of the queue, maybe because there is cycle there.

If I insert a `currentTaskPackage = nullptr;` after line 512 and force the deleter task to be the last thing holding the queue by adding tome sleeps in test, then I see the expected exception.

jkelling replied:
In the old PR https://github.com/alpaka-group/alpaka/pull/1727 which would only fix the warning, I modified the test and `ConcurrentExecPool` to catch this problem, so the can call that version "red".

jkelling replied:
Apart from support for boost::fiber, this PR is ready.

@alpaka-group/alpaka-maintainers  please review.

jkelling committed:
memBufAsyncConstTest: delay to force BufCpu::deleter to delete queue

ConcurrentExecPool<non-yielding>::concurrentExecFn: limits lifetime of
currentTaskPackage
This makes memBufAsyncConstTest fail due to the captured queue causing a
thread to self-join

jkelling committed:
QueueGenericThreadsNonBlocking: Add async task completion on destruction

Add ConcurrentExecPool::detach()
Add delete QueueGenericThreadsNonBlockingImpl::(move,copy)

Add trait IsThisThread

Add header core(ThreadTraits.hpp
Add imply for std::thread and boost::fiber

Add detach in case isThisThread in joinAll
Add m_bDetachFlag atomic to communicate detach state
<yield=true>: enshure only one thread dletes by atomic exchange
<yield=false>: avoid deadlock and write-after free by unlocking wakeup
mutex

Fixup QueueGenericThreadsNonBlocking: NSDMI for all atomic, ...

... one decl per line

jkelling commented:
It appears we need to introduce a trait for `std::this_thread::get_id()` in the check `std::this_thread::get_id() != concurrentExec.get_id()`, because at the moment test are failing because this code break boost::fiber. On the other hand there is https://github.com/alpaka-group/alpaka/pull/1718 .

@j-stephan Should other thread type remain supported, or will this be merged soon?

jkelling commented:
I added a new trait `IsThisThread`, having this is more clean even with boost::fiber gone and the boost::fiber implementation of removes the noise of failing boost::fiber tests.

There are two failing runs on windows: [windows_cl-2019_release](https://github.com/alpaka-group/alpaka/runs/6591158601?check_suite_focus=true) and [windows_cl-2022_release](https://github.com/alpaka-group/alpaka/runs/6591158782?check_suite_focus=true) in both cases a similar set of test is failing and all failing test print with `***Exception: `, but no further information. The debug variants of these are succeeding. Could there an issue with using `std::this_thread::get_id()` in msvc? @bernhardmgruber do you have an idea?

jkelling commented:
Deleting a not-joined `std::thread` should throw an exception and there is no difference if a thread destructs it own handle. So, the thread that `isThisThread` and which cannot be joined must be detached instead. Curious, that I no test caught this, except those msvc release builds, at lest they are passing now with this in place.

I also added some code to make `ConcurrentExecPool::detach()` also work when there is more than one worker thread.

BenjaminW3 commented on include/alpaka/core/ConcurrentExecPool.hpp:
```diff
@@ -367,7 +385,8 @@ namespace alpaka::core
             ThreadSafeQueue<std::shared_ptr<ITaskPkg>> m_qTasks;
             std::atomic<std::uint32_t> m_numActiveTasks;
-            std::atomic<bool> m_bShutdownFlag;
+            std::atomic<bool> m_bShutdownFlag, m_bDetachedFlag = false;
```
If NSDMI works the same way variable initialization works in other places, then this does not do what you think.
Only `m_bDetachedFlag` is set to `false` and `m_bShutdownFlag` is default initialized.
Please split this initialization in multiple lines.

jkelling replied:
It does work the same way and intended to get this behaviour. So, the two-line version would be:
```c++
std::atomic<bool> m_bShutdownFlag;
std::atomic<bool> m_bDetachedFlag = false;
```
`m_bShutdownFlag` was there before and is initialized in the c'tor. I did not want to touch code I do not need to change but also did not want to use old-style initialization for the new member.

I can see how this inconsistency obfuscates the intent, so I guess it is best to update the trivial initializations of the pre-existing members.

BenjaminW3 commented on include/alpaka/core/ConcurrentExecPool.hpp:
```diff
@@ -551,7 +594,9 @@ namespace alpaka::core
             TMutex m_mtxWakeup;
             TCondVar m_cvWakeup;
-            std::atomic<bool> m_bShutdownFlag;
+            std::atomic<bool> m_bShutdownFlag, m_bDetachedFlag = false;
```
same here

jkelling commented:
It does work the same way and intended to get this behaviour. So, the two-line version would be:
```c++
std::atomic<bool> m_bShutdownFlag;
std::atomic<bool> m_bDetachedFlag = false;
```
`m_bShutdownFlag` was there before and is initialized in the c'tor. I did not want to touch code I do not need to change but also did not want to use old-style initialization for the new member.

I can see how this inconsistency obfuscates the intent, so I guess it is best to update the trivial initializations of the pre-existing members.

On 26 May 2022 08:26:13 CEST, Benjamin Worpitz ***@***.***> wrote:
***@***.*** commented on this pull request.
>
>
>
>> @@ -367,7 +385,8 @@ namespace alpaka::core
>             std::vector<TConcurrentExec> m_vConcurrentExecs;
>             ThreadSafeQueue<std::shared_ptr<ITaskPkg>> m_qTasks;
>             std::atomic<std::uint32_t> m_numActiveTasks;
>-            std::atomic<bool> m_bShutdownFlag;
>+            std::atomic<bool> m_bShutdownFlag, m_bDetachedFlag = false;
>
>If NSDMI works the same way variable initialization works in other places, then this does not do what you think.
>Only `m_bDetachedFlag` is set to `false` and `m_bShutdownFlag` is default initialized.
>Please split this initialization in multiple lines.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/alpaka-group/alpaka/pull/1728#pullrequestreview-985778315
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

jkelling commented:
Three CI runs for the new version containing only some style changes (https://github.com/alpaka-group/alpaka/pull/1728/commits/5f4dd68de50a060ef063be95ce7bcfee3dcf7ecc). 

In two cases gcc got killed:
*  [linux_gcc-8_debug ](https://github.com/alpaka-group/alpaka/runs/6607231273?check_suite_focus=true) (In this case the GUI view does not load the output, but the raw logs can be downloaded.)
   ```
   2022-05-26T10:15:38.9305827Z g++: fatal error: Killed signal terminated program cc1plus
   2022-05-26T10:15:38.9573880Z compilation terminated.
   2022-05-26T10:15:38.9596836Z make[2]: *** [test/unit/kernel/CMakeFiles/kernelTest.dir/build.make:147: test/unit/kernel/CMakeFiles/kernelTest.dir/src/KernelWithOmpSchedule.cpp.o] Error 1
   ```
* [linux_gcc-9_debug_c++17](https://github.com/alpaka-group/alpaka/runs/6615775997?check_suite_focus=true)
   ```
   g++: fatal error: Killed signal terminated program cc1plus
   compilation terminated.
   make[2]: *** [test/unit/kernel/CMakeFiles/kernelTest.dir/build.make:146: test/unit/kernel/CMakeFiles/kernelTest.dir   /src/KernelWithOmpSchedule.cpp.o] Error 1
   ```

Is it possible,that here the container went OoM?

One case is a bit concerning, because it is a SIGSEGV in `memViewTest` (at the end of the program, after all checks succeeded), but [linux_clang-5_debug_c++17](https://github.com/alpaka-group/alpaka/runs/6608821015?check_suite_focus=true#step:5:7584). I cannot reproduce this locally, nor in the CI it seems. This may be related to the changes in this PR, or may be something else... . Any suggestions?

jkelling commented:
> Three CI runs for the new version containing only some style changes ([5f4dd68](https://github.com/alpaka-group/alpaka/commit/5f4dd68de50a060ef063be95ce7bcfee3dcf7ecc)).

Btw.: by "three CI runs" I mean three different triggered actions, each of those had only one check fail, i.e. each failed check succeeded in the other two.

jkelling commented:
With the queues getting detached, there is a chance, that the worker thread of a queue runs past the end of `main()`, invoking UB. Such a thing may have caused the SIGSEGV in that clang check. (The current state of develop has the same problem, when a queue's worker thread gets leaked by the `BufCpu::deleter`)

If we want the non-blocking queue not to block in its c'tor, we need some other place to make sure the detached queue can be caught before program termination. The device may be a place to do it. The device's d'tor waiting for all task on the device to finish seems reasonable to me. The device is already connected to its queues, though the current list of `weak_ptr` would not do. A destructed queue would need to register a "waiter" that makes the d'tor wait for the queue (`ConcurrentlExecPool`) to actually destruct. This does add some complexity...

jkelling committed:
Dev{CpuOacc,Omp5}: Add QueueRegistry to unify code

jkelling committed:
QueueGenericThreadsNonBlocking: Add register cleaner with device

QueueRegistry: Fix spQueue decl style
ConcurrentExecPool: Add takeDetachHandle()

jkelling committed:
queueTest: Add nonBlockingQueueShouldNotRunPastProgramTermination

jkelling commented:
Two groups checks are failing now, but consistently:
1. clang <=8 with libc++: In some tests an exception is thrown that indicates, that an attempt made to lock a mutex which was already destructed, or not allocated. I wan reproduce this locally using clang 8 with libc++ and having an OMP2 backend enabled. The error requires these three things in combination to occur. Considering, that this requires older version of clang(libc++ (<= 8) and it does not occur with newer versions, neither with libstdc++, this is probably an old bug in libc++ which got fixed by now. As this affects an older compiler version, I prefer to not investigate further and either introduce a work around that make the queue wait in its c'tor or deprecate libc++ <= 8 . @alpaka-group/alpaka-maintainers do we have a defined support level for libc++?
2. With the non-cuda msvc debug builds, ~~the `fenceTest` is~~ some tests (not sure where I saw only the one) failing (says "segfault"), while tests pass in the release build, at least for the latest run, this may not be consistent. Does anyone with access to msvc have any idea what is going on there? This does look a bit like like the clang+libc++<=8-issue, only with a less specific exception.

fwyzard commented on test/unit/queue/src/QueueTest.cpp:
```diff
@@ -171,3 +172,34 @@ TEMPLATE_LIST_TEST_CASE("queueShouldNotExecuteTasksInParallel", "[queue]", TestQ
+//! not synchronized somewhere. This test has no assertion, because the tested
+//! behviour is outside of any function scope.
+//!
+//! A thread running past program termination invoked undefined behavior, which
```
```suggestion
//! A thread running past program termination invokes undefined behavior, which
```

fwyzard commented on test/unit/queue/src/QueueTest.cpp:
```diff
@@ -171,3 +172,34 @@ TEMPLATE_LIST_TEST_CASE("queueShouldNotExecuteTasksInParallel", "[queue]", TestQ
     secondTaskFinishedFuture.get();
 }
+
+//! This task launches a long task in a non-blocking queue and destroys the
+//! queue before the task is finished. The run time of the task is a bit longer
+//! than all other tests here, so the task would run past program termination if
+//! not synchronized somewhere. This test has no assertion, because the tested
+//! behviour is outside of any function scope.
+//!
+//! A thread running past program termination invoked undefined behavior, which
+//! occasionally leads to crashes after termination. I.e. this test may crash
+//! after catch2 reported success.
```
Would a catch2 check at the end of the long task be useful in marking the "undefined behaviour" as an explicit failure ?

jkelling replied:
No, because that check would just never happen in the failure scenario: The thread runs and either causes some trouble, because it touches memory or calls a function related to something already cleaned-up after `main()` or has no effect and at some point just gets killed by the OS when parent process terminates. So, the thread would not get to that check. If we were to wait for the check, then we eliminate the failure scenario, by effectively joining that thread.

It would be better if we could check for some defined behavior instead of checking for nothing undefined happening. For example, if in alpaka all queues devices were required to synchronize with and destroy their attached queues in their destructor, we could check that. But this is not a thing that alpaka devices do.

fwyzard commented on test/unit/mem/buf/src/BufTest.cpp:
```diff
@@ -162,27 +162,35 @@ TEMPLATE_LIST_TEST_CASE("memBufConstTest", "[memBuf]", alpaka::test::TestAccs)
+        // delay the end end of the queue to push the buffer deletetion task
+        // after all local refs to queue have been dropped
+        alpaka::enqueue(queue, []() { std::this_thread::sleep_for(std::chrono::microseconds(1000)); });
+        [](auto) {}(std::move(queue));
```
What is the usefulness of this lambda ? It is executed immediately.

jkelling replied:
The lambda destroys the `queue` handle in the local scope, by moving it into another scope, where it immediately goes out of scope. This there to remove the local `queue` handle before the buffer is destructed (opposite to the normal destruction order), to leave the last handle inside the buffer or its deleter.

fwyzard replied:
I see.

But why do you need a lambda ?
Wouldn't moving from the queue in a nested scope be sufficient ?
```suggestion
        {
          // ...
          auto unused = std::move(queue);
        }
```

jkelling replied:
Might do (apart from the `}` there, but would this not raise an "unused" warning?

fwyzard replied:
I would argue that it is equally unused in both cases... 
But if it causes a warning, I guess
```suggestion
        {
          auto unused [[maybe_unused]] = std::move(queue);
        }
```
should do :-/

jkelling replied:
But in case of the lambda the missing parameter name tells the compiler, that it is expected to be unused.

fwyzard commented on test/unit/mem/buf/src/BufTest.cpp:
```diff
@@ -162,27 +162,35 @@ TEMPLATE_LIST_TEST_CASE("memBufConstTest", "[memBuf]", alpaka::test::TestAccs)
 
-    // the buffer will queue the deallocation of the memory when it goes out of scope,
-    // and extend the lifetime of the queue until all memory operations have completed.
+    std::this_thread::sleep_for(std::chrono::microseconds(1200));
```
Why is this needed ?

jkelling replied:
This makes sure, the test case does not end before the buffer's deleter ran, so any exceptions thrown during it will show up there. At this point the queue is detached (both on `develop` and in this PR), so it does not synchronize with the end of this scope.

fwyzard commented on include/alpaka/queue/QueueGenericThreadsNonBlocking.hpp:
```diff
@@ -57,12 +57,30 @@ namespace alpaka
+                                s = s->takeDetachHandle();
+                            }
+                        });
+                    m_workerThread->detach(std::move(m_workerThread));
```
Is this legal ?
`detach()` is invoked on `m_workerThread` _after_ is has been moved from.

jkelling replied:
Yes.

All arguments for the function are obtained first (i.e. `this` for `detach()` and the rvalue smart pointer) in no specific order, but before the function `detach` is invoked. `std::move` only casts the reference to the object into an rvalue reference (basically it only affects overload resolution), it does not modify the object. The actual change to `m_workerThread`, which puts it into a moved-from state, happens inside `detach`, in for of the move-assignment.

jkelling replied:
Also note, that `detach()` is not invoked on`m_workerThread`, which is a `shared_ptr<>` (in the latest version), but on `*m_workerThread`, which is a `ConcurrentExecPool`.

jkelling replied:
> in no specific order,

This is actually not correct: the `operator->` constitutes a sequence point, so `this` is obtained first, then the other function arguments are evaluated and assigned (those in no specific order, but there is only the one here). This does not really matter as `detach()` takes an rvalue reference and thus the rest of my post applies. If it would take the argument by value (which be wrong from a API-semantic point of vierw), then a move assignment would happen during function argument assignment, but then the sequence point would still make it behave in the same way.

fwyzard replied:
Thanks for the explanation, I tend to agree that the actual "moving from" happens inside the function body, after the function call.

Just on one point:

> the `operator->` constitutes a sequence point,

I cannot find any reference to this being the case, could you point to it ?

jkelling replied:
Hmm, maybe there is actually no sequence defined.

[Here](https://en.cppreference.com/w/cpp/language/eval_order) under "Rules", rules 2 or 14 could apply somehow, but 2 sequences only the value computations, so the side-effect of a move-assignment in the function-call expression could still go anywhere. I do not know, if, in 14, the "expression that names the function" includes the object and member access, or only anything that is where the function name goes, like some expression evaluation to a function pointer.

fwyzard replied:
**disclaimer**: *this is a contrived example and has nothing to do with the code in the PR*.

But I was curious, so I've tried to come up with a simpler example where the order of evaluation would actually have observable consequences: https://godbolt.org/z/j6c1GK8Tj .

I _think_ the order of evaluation in `p->print(move_from_and_return(p))` is the same as in `m_workerThread->detach(std::move(m_workerThread))`.

In the contrived example, the argument to the function _does_ "move from" the `shared_ptr`  _before_ the method is called, resulting in the call appling to a `nullptr`, and leading to a crash.

jkelling replied:
I also played with some example, but only with different ways of assigning the arguments (e.g. by-value instead of by-rvalue-ref), not with calling a function that has side-effects (`std::move` is only a cast). Intuitively, I think that the evaluation of the call to `move_from_and_return()` might even be guaranteed to happen before the outer parts of the expression, not just be called first depending on the compiler's whim.

I think we clarified, that `->` does not imply any sequence of evaluation here. I.e., if `detach` would take a value, than the compiler may do the move-assignment of the parameter or the dereference at `->` first as it chooses ... Or it maybe, since C++17, it is compelled to evaluate the part identifying the function first before assigning its argument, not sure.

jkelling replied:
> Intuitively, I think that the evaluation of the call to `move_from_and_return()` might even be guaranteed to happen before the outer parts of the expression

Although, I do not see any of the rules implying this, so I guess the order in your example is probably undefined.

fwyzard reviewed (commented):
Could you make two separate PRs, one with only the changes to `include/alpaka/mem/buf/BufCpu.hpp`, and one with the semantic changes to `QueueGenericThreadsNonBlocking` ?

jkelling commented:
> Could you make two separate PRs, one with only the changes to `include/alpaka/mem/buf/BufCpu.hpp`, and one with the semantic changes to `QueueGenericThreadsNonBlocking` ?

No, because the change to `BufCpu` only makes it broken in a different way: Currently the worker thread of the queue is leaked,. The queue itself remains referenced by the thread, so is technically not leaked, which is why the address sanitizer check passes. The change makes the buffer's memory occasionally leak, making that test fail.

jkelling commented:
Regarding these failing checks with clang + libc++ <=8 (and maybe likewise the msvc checks):

The exception from the mutex lock traces back into the worker thread of the queue. I.e. it looks like the mutex was destructed while the worker thread is still running. This should not be possible because the mutex is a member of `ConcurrentExecPool` and is just only destructed at the end of `~ConcurrentlExecPool`, which joins all worker threads, so does not reach the end while threads are running. The only exception to that is, when a worker thread destruct the `ConcurrentExecPool` in which case the thread is not joined but detached, so keeps running. However: the worker thread destructs the pool only in a [return statement](https://github.com/alpaka-group/alpaka/pull/1728/files#diff-1e7ed7f50e4605ed7f5e84260029cb16b4a8caf21e02e3831ebc116eef9e8448R558-R563), so it cannot be running afterwards. Ultimately, this could only be explained by a memory corruption in the code, if there is not weird compiler bug.

A memory corruption, and the fact, that an OpenMP backend must be active for this to happen, causes the bad feeling, that this might be related to https://github.com/alpaka-group/alpaka/issues/1676 : This PR adds some interesting ways of handling `std::threads` spawned inside an OpenMP environment, which is not required to be thread-safe and is allowed to assume, that all threads were spawned by the OpenMP runtime...

jkelling committed:
QueueRegistry: make cleanupList static inside registerCleanup

jkelling commented:
So, the rabbit hole is even deeper:

With the clang8 + libc++ + OMP2 setup, I found, that the invalid mutex lock occurred in a worker thread, that actually had destructed the Concurrent exec pool itself. A backtrace from that point of destruction lead through a `TaskPkg` holding an `EventThreadGeneric` which in turn owns a `Dev`. That `Dev` was apparently the last handle in this situation, so the destruction of the task package, which also destructed the event (which also must have been the last handle, so no one outside seemed to care about the event anymore), destructed the `ConcurentExecPool` (i.e. the involved queue must have been gone before that).

When I added the `m_cleanupList` in the device to catch detached worker `ConcurrentExecPool`s, I assumed, that devices would be sure to outlive their queues and would not get destructed inside queues ... alpaka Event seems to have a different opinion. I pushed a commit that makes the `cleanupList` a static variable inside the `Dev::registerCleanup()` function, so  it is no longer attached to a device instance. Most programs probably only hold one `Dev` per physical device anyway, so having a global mutex per type instead of one per instance, should not change anything performance-vise. An issue could be, that the list only grows when queues are destroyed but never shrinks.So if queue are used in a fire-and forget fashion, this can cause memory issues...

jkelling commented:
@alpaka-group/alpaka-maintainers Is there a reason for there being no hierarchy between devices an queues? Each queue seems to be attached to exactly one device over its life time and the `shared_ptr` connection, ensures, that the device's lifetime cover that of the queue. Why does the device not just own the queue? We could just have an `std::unique_ptr<IQueueType>` in the device and a plain pointer back the queue... .

jkelling commented:
The last CI run went through without failures. I rebased and cleaned up the branch.

fwyzard commented on include/alpaka/dev/common/QueueRegistry.hpp:
```diff
@@ -0,0 +1,102 @@
+{
+    namespace detail
+    {
+        //! The CPU device implementation.
```
is the comment still correct ?

jkelling replied:
It is indeed not correct anymore.

fwyzard commented:
@tonydp03 could you run the CPU-only benchmark of [pixeltrack-standalone](https://github.com/cms-patatrack/pixeltrack-standalone/pulls/fwyzard) with the `HEAD` of the alpaka `devel` branch, with and without this PR ?

fwyzard commented:
> @tonydp03 could you run the CPU-only benchmark of [pixeltrack-standalone](https://github.com/cms-patatrack/pixeltrack-standalone/pulls/fwyzard) with the `HEAD` of the alpaka `devel` branch, with and without this PR ?

the results from @tonydp03 do not show any impact on the performance:

> ![image](https://user-images.githubusercontent.com/4069793/171673883-c9a6162b-25af-4d21-b303-869c3d229f65.png)

psychocoderHPC commented:
I will not be able to check this PR before mid of the next week. sry

psychocoderHPC commented on include/alpaka/core/ConcurrentExecPool.hpp:
```diff
@@ -503,12 +503,14 @@ namespace alpaka::core
-
-                    // Use popTask so we only ever have one reference to the ITaskPkg
-                    if(popTask(currentTaskPackage))
                     {
```
please add a comment on why this code is in its own code block. I assume it has something to do with guaranteeing when ` currentTaskPackage` is destroyed. The comment should avoid that someone is removing the paranteeses.

psychocoderHPC commented on test/unit/mem/buf/src/BufTest.cpp:
```diff
@@ -162,27 +162,38 @@ TEMPLATE_LIST_TEST_CASE("memBufConstTest", "[memBuf]", alpaka::test::TestAccs)
+
+        // the buffer will queue the deallocation of the memory when it goes out of scope,
+        // and extend the lifetime of the queue until all memory operations have completed.
+        // delay the end end of the queue to push the buffer deletetion task
```
`end end`?

jkelling commented:
Force-pushed nothing to restart pipeline `linux_gcc-9_debug_c++17 ` failed because `g++: fatal error: Killed signal terminated program cc1plus`.

jkelling commented:
The CI is showing an occasional timeout in `queueTest`. It seems to always happen in `QueueUniformCudaHipRt` (with either value for `TBlocking`). I am investigating.

jkelling commented:
This PR is ready now. The `queueTest` timeout is a pre-existing bug, reported in #1741 .

@psychocoderHPC please merge.
